scripts/build_bus_regions.py

```diff
@@ -256,6 +256,22 @@ def get_gadm_shape(
         crs=country_shapes.crs,
     ).dropna(axis="index", subset=["geometry"])
 
+    if snakemake.params.alternative_clustering:
+        # determine isolated buses
+        n.determine_network_topology()
+        non_isolated_buses = n.buses.duplicated(subset=["sub_network"], keep=False)
+        isolated_buses = n.buses[~non_isolated_buses].index
+        # drop isolated buses
+        onshore_regions = onshore_regions[~onshore_regions.name.isin(isolated_buses)]
+        # drop duplicates based on shape_id
+        onshore_regions = onshore_regions.drop_duplicates("shape_id")
```

Comment:
```suggestion
        non_isolated_regions = onshore_regions[~onshore_regions.name.isin(isolated_buses)]
        isolated_regions = onshore_regions[onshore_regions.name.isin(isolated_buses)]
        
        # Combine regions while prioritizing non-isolated ones
        onshore_regions = (
            pd.concat([non_isolated_regions, isolated_regions])
            .drop_duplicates("shape_id", keep="first")
        )
```

Could you check if this works? this also allows isolated nodes to be kept if necessary. It builds on your spirit.
The logic is that non_isolated_nodes have priority over isolated ones and isolated may be kept if no other bus is available there.
Could you check if it works?

Reply:
Thank you @davide-f. I checked and this works! I think it could add some stablility.
So this code prefers non-isolated bus but isolated one could represent a region if the region got only isolated ones.

Reply:
The edited code be like

```python

    if snakemake.params.alternative_clustering:
        # determine isolated buses
        n.determine_network_topology()
        non_isolated_buses = n.buses.duplicated(subset=["sub_network"], keep=False)
        isolated_buses = n.buses[~non_isolated_buses].index
        non_isolated_regions = onshore_regions[~onshore_regions.name.isin(isolated_buses)]
        isolated_regions = onshore_regions[onshore_regions.name.isin(isolated_buses)]
        
        # Combine regions while prioritizing non-isolated ones
        onshore_regions = (
            pd.concat([non_isolated_regions, isolated_regions])
            .drop_duplicates("shape_id", keep="first")
        )
        
        if len(onshore_regions) < len(gadm_country):
            logger.warning(
                f"The number of remaining of buses are less than the number of administrative clusters suggested!"
            )
```